File: java/freeman/command/undo/RemoteControl.java
public class RemoteControl{
    Command[] onCommands;
    Command[] offCommands;
    Command lastCommand;
    public RemoteControl(){
        onCommands=new Command[5];
        offCommands=new Command[5];
        Command noCommand=new NoCommand();
        for (int i=0;i<5;i++){
            onCommands[i]=noCommand;
            offCommands[i]=noCommand;
        }
        lastCommand=noCommand;
    }
    public void setCommand(int slot,Command onCommand,Command offCommand){
        onCommands[slot]=onCommand;
        offCommands[slot]=offCommand;
    }
    public void onButtonWasPressed(int slot){
        System.out.println("Somebody pressed on button in slot "+slot+"(operation:"+onCommands[slot].getClass().getName()+")");
        onCommands[slot].execute();
        lastCommand=onCommands[slot];
    }
    public void offButtonWasPressed(int slot){
        System.out.println("Somebody pressed off button in slot "+slot+"(operation:"+offCommands[slot].getClass().getName()+")");
        offCommands[slot].execute();
        lastCommand=offCommands[slot];
    }
    public void undoButtonWasPressed(){
        System.out.println("undo operataion:"+lastCommand.getClass().getName());
        lastCommand.undo();
    }
    public String toString(){
        StringBuffer stringBuff=new StringBuffer();
        stringBuff.append("----------------Remote Control------------------\n");
        for (int i=0;i<onCommands.length;i++){
            stringBuff.append("[slot "+i+"]"+onCommands[i].getClass().getName()+" "+offCommands[i].getClass().getName()+"\n");
        }
        return stringBuff.toString();
    }
}

File: php/oop-study/lesson3/Macro/RemoteControl.php
<?php
class RemoteControl{	
	private $onCommands=array();
    private $offCommands=array();
	private $lastCommand;//Последняя операция
    public function __construct(){
        $noCommand=new NoCommand();
        for ($i=0;$i<3;$i++){//Инициализируем массив пустыми командами
            $this->onCommands[i]=$noCommand;
            $this->offCommands[i]=$noCommand;
        }
		$this->lastCommand=$noCommand;
    }
    public function setCommand($slot,Command $onCommand,Command $offCommand){
        $this->onCommands[$slot]=$onCommand;
        $this->offCommands[$slot]=$offCommand;
    }
    public function onButtonWasPressed($slot){
        echo "Somebody pressed on button in slot ".$slot." (command:<i>".get_class($this->onCommands[$slot])."</i>)"."<br>";
        $this->onCommands[$slot]->execute();
		//Меняем последнюю выполненную команду
		$this->lastCommand=$this->onCommands[$slot];
    }
    public function offButtonWasPressed($slot){
        echo "Somebody pressed off button in slot ".$slot." (command:<i>".get_class($this->offCommands[$slot])."</i>)"."<br>";
        $this->offCommands[$slot]->execute();
		//Меняем последнюю выполненную команду
		$this->lastCommand=$this->offCommands[$slot];
    }   
	public function undoButtonWasPressed(){//Отменяем последнюю выполненную команду
		echo "----------------------------------------------------------<br>";
        echo "<b>undo operation:</b> <i>".get_class($this->lastCommand)."</i><br>";
        $this->lastCommand->undo();
		echo "----------------------------------------------------------<br>";
    } 	
}
?>